### PR TITLE
Adds workflow approval list and details

### PR DESCRIPTION
**Pull request by mabashian**
##### SUMMARY
Meta issue https://github.com/ansible/awx/issues/8273
Addresses #8373 
Sort of addresses https://github.com/ansible/awx/issues/7112 (the rest of it captured in https://github.com/ansible/awx/issues/8374)
Does not include https://github.com/ansible/awx/issues/8372

This PR adds a new list `/workflow_approvals` along with a details view for items in that list.

The entry point for this list is here:

Screen Shot 2020-10-12 at 2 36 43 PM

Since this list does not exist in the old UI I figured this was as good a spot as any.  We can move it if we need to.

The list itself is pretty straightforward but here's an explanation of the columns:

Screen Shot 2020-10-12 at 2 45 12 PM

Screen Shot 2020-10-12 at 2 47 18 PM

Screen Shot 2020-10-12 at 2 50 26 PM

Some examples of tooltip on status:

Screen Shot 2020-10-12 at 2 51 00 PM
Screen Shot 2020-10-12 at 2 51 06 PM

Screen Shot 2020-10-12 at 2 54 38 PM

Here's a look at the details view for a pending approval:

Screen Shot 2020-10-12 at 2 55 15 PM

a lot of the same information from the list row is shown here.

Here's a look at the details view for an approved approval:

Screen Shot 2020-10-12 at 2 56 11 PM

The action buttons/rbac should be the same on the details view as they are on the list.

**Websockets**

OK so here's how websockets work on this list.  This list is subscribed to the jobs -> status_changed channel just like the jobs list.  We're only interested in jobs with the type: `workflow_approval` though.  The response to events that match our criteria is pretty straightforward: we refresh the list contents.  This is throttled but that throttle is currently set to just 1 second.  Unlike the jobs list, I'm not anticipating a tremendous amount of activity on this list in a short amount of time but if we need to increase that throttle threshold then we can.  Here's when we refresh the list:

1) When the job in question _is not_ in our list _and_ it's status is `pending`.
2) When the job in question _is_ in our list _and_ it's status is not in `['new', 'pending', 'waiting', 'running']` (i.e. the job finished).

^^ Scenario 2 creates an interesting challenge when the user that's viewing the list is also the user that is acting on the row.  When the user clicks approve/deny and the API responds with a 204 there's a bit of a delay before the socket message comes through and tells us that the job has moved to a completed status.  During that time, I hide the action buttons via a boolean on the list row state.

**Websockets are where I spent a majority of my time on this feature and I'm still not sure that it's quite right.  I'm definitely open to alternative approaches to keeping this list up to date using the data that we get from the job status channel.**

Unfortunately, we have to hit the API to retrieve the row data as the socket message doesn't have much more than the status and the job id.  There could be a solution where we fetch specific rows instead of refetching all of them.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

### CODE REVIEW

**Comment by wenottingham** [comment]
Looks generally good. I do think we need to move the list - workflow approvals are the same conceptual things as job runs, and so should be in that area.

**Comment by unlikelyzero** [comment]
@tiagodread note: This is net new and not absolutely critical for this week.

**Comment by mabashian** [comment]
OK, made the following updates based on feedback from Taufique and Bill.

1) Moved Workflow Approvals out of Administration and in to Views
2) Moved approve/deny buttons from the individual list row items up to the list action level (alongside delete).  This means multiple rows can be approved/denied at the same time.
3) Updated the UX on the status labels to match the mockups: https://tower-mockups.testing.ansible.com/patternfly/workflow-approvals/wf-approvals/

**Review by jakemcdermott — COMMENTED** [review]
- List item, approval button, new view, and tests, etc. LGTM
- let's double-check w/ @keithjgrant on the websocket hook usage

File: awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx
Comment (keithjgrant) [inline awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx:9]: We have a generic status label component at `components/StatusLabel`. Do you think it could meet your needs here instead of a new component?

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalDetail/WorkflowApprovalDetail.jsx
Comment (AlexSCorey) [inline awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalDetail/WorkflowApprovalDetail.jsx:231]: You could combine these 2 AlertModals, and would you also only need to have 1 useDismissableError above, which would be something like
```const {error, dismissError} = useDismissableError(approveError || denyError)```
and then the alertModals would be
```<AlertModal
          isOpen={error}
          variant="error"
          title={i18n._(t`Error!`)}
          onClose={dismissError}
        >
         approveError ?  {i18n._(t`Failed to approve workflow approval.`)} 
            : {i18n._(t`Failed to deny workflow approval.`)}
          <ErrorDetail error={approveError} />
        </AlertModal>
```

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.jsx
Comment (AlexSCorey) [inline awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.jsx]: Is this a note for your current work that you forgot to remove, or is this future work?

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.test.jsx
Comment (AlexSCorey) [inline awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.test.jsx]: Might consider putting this inside of a data.json file to reduce the # of lines of code in this file

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/useWsWorkflowApprovals.js
Comment (AlexSCorey) [inline awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/useWsWorkflowApprovals.js:1]: was does the Ws in `useWsWorkflowApprovals` stand for?

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.jsx
Comment (keithjgrant) [reply awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.jsx]: that's an old TODO of mine, copied over from somewhere else I think

**Comment by tiagodread** [comment]
Hey @mabashian !
- I tried to delete a workflow approval before approving or deny and I got this result (on the list or details):

approval

- The simple search using name doesn’t work:
search

- If i delete the job the workflow approval associated with that job show `undefined` for the job name and if I click on it I got 404:

deleted

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalListApproveButton.jsx
Comment (unlikelyzero) [inline awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalListApproveButton.jsx:54]: do aria-labels need to be translated?

**Comment by unlikelyzero** [comment]
We have tests pending the merge of this PR. Please let us know when it's ready for final merge

**Comment by tiagodread** [comment]
@mabashian can you add some uniques ids for the data on datails tab specially for the status

File: awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx
Comment (mabashian) [reply awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx:9]: 👍 will check that out

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.test.jsx
Comment (mabashian) [reply awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.test.jsx]: 👍

File: awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/useWsWorkflowApprovals.js
Comment (mabashian) [reply awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/useWsWorkflowApprovals.js:1]: That's the naming pattern we have for all of our websocket hooks.  They all start with `useWs` and then we concat the resource on the end.

File: awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx
Comment (mabashian) [reply awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx:9]: Yea there's definitely some overlap here.  If we want to keep the same UX then I'd have to refactor this component quite a bit with several conditionals.  I'm a little reluctant to do that given how clean this component (StatusLabel) is.

**Comment by mabashian** [comment]
OK @tiagodread @unlikelyzero I think this is ready for your eyes again.  I've added data-cy attributes to all the details: https://github.com/ansible/awx/pull/8375/files#diff-eeea31a09935ea3101ae734da15ca14d933119b4bb0819194fb9ff24a074268cR85-R183

**Comment by unlikelyzero** [comment]
@mabashian hey, we're not longer using data-cy attributes but trying to use the dataouia attributes instead. Is there enough to glean from this PR?

https://github.com/ansible/awx/pull/8472#discussion_r517504241

**Comment by tiagodread** [comment]
Hey @mabashian looks like the error `- I tried to delete a workflow approval before approving or deny and I got this result (on the list or details)` was fixed for the list, but, i can try to delete a pending workflow approval in the details page:

wfaproval

**Comment by tiagodread** [comment]
@mabashian awesome job!!!

Tests covered by a mix of manual and automated test cases in here https://github.com/ansible/tower-qa/pull/5656

**Review by tiagodread — APPROVED** [review]

**Comment by mabashian** [comment]
@keithjgrant @jakemcdermott @AlexSCorey any other thoughts on this one before merge?

**Review by jakemcdermott — APPROVED** [review]

**Pull request merged**
